storage/engine: MVCC Metamorphic test suite, first phase

itsbilal opened the pull request:
This PR adds a new test-only sub-package to engine, metamorphic,
which has one test, TestMeta, that generates and runs random MVCC
operations on rocksdb and pebble instances with default settings.

Future additions to this test suite could include:

- [x]  A "check" mode that takes an output file as input, parses it, runs the operations in that sequence, and compares output strings.
- [ ]  Diffing test output between rocksdb and pebble and failing if there's a difference
- [ ]  Adding support for more operations
- [ ]  Adding a "restart" operation that closes the engine and restarts a different kind of engine in the store directory, then confirming operations after that point generate the same output.

First-but-biggest part of #43762 .

Release note: None

Commit:
storage/engine: MVCC Metamorphic test suite, first phase

This PR adds a new test-only sub-package to engine, metamorphic,
which has one test, TestMeta, that generates and runs random MVCC
operations on rocksdb and pebble instances with default settings.

Future additions to this test suite could include:

 - A "check" mode that takes an output file as input, parses it,
   runs the operations in that sequence, and compares output
   strings.
 - Diffing test output between rocksdb and pebble and failing if
   there's a difference
 - Adding support for more operations
 - Adding a "restart" operation that closes the engine and
   restarts a different kind of engine in the store directory,
   then confirming operations after that point generate the
   same output.

First-but-biggest part of #43762 .

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44458)
<!-- Reviewable:end -->

itsbilal commented:
I'll designate this as WIP for now until I finish the ability to parse/check an output file later today. Until that happens, some of the parsing code is unused and buggier than I'd like.

Commit:
engine/metamorphic: Add ability to re-check past runs with output files

When run with the --check <filename> flag, TestMeta will parse
the specified output file from a past TestMeta run for operations.
It will run them in that order, and check for matching output
strings. If there's a mismatch, the test will throw an error.

Will be useful for when this test suite will try swapping engine
types after a restart and will verify if the output stays the same.

Release note: None.

itsbilal commented:
Fixed bugs in the parsing paths, `TestMeta` when re-run with its own output files now passes. Ready for review again.

petermattis reviewed (commented):
First pass comments that jumped out at me. I'll continue to review this later today.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-:-LzmmS0w5-f8uxg92_gW:bdqn9sg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @petermattis, and @sumeerbhola)

___
*[pkg/storage/engine/metamorphic/generator.go, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzmlExm6KeNUqc9bOCu:-LzmlExm6KeNUqc9bOCv:bfha9y7) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8bf979a8ba42a5140800da6f3a0c05114d207e/pkg/storage/engine/metamorphic/generator.go#L35)):*
> ```Go
> 	engine      engine.Engine
> 	tsGenerator tsGenerator
> 	managers    map[operandType]operandManager
> ```

I suspect there is a level of abstraction here that isn't serving much purpose. Given that there are a fixed number of "*managers", rather than a map you could embed the concrete types here. There are a few places where you're using a variable as a key to `managers`, but my suspicion is that is due to the operand abstraction which I'm also questioning.
___
*[pkg/storage/engine/metamorphic/operands.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-Lzmlz4zEP89aGWCO0vI:-Lzmlz4zEP89aGWCO0vJ:b-nxipi5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8bf979a8ba42a5140800da6f3a0c05114d207e/pkg/storage/engine/metamorphic/operands.go#L30)):*
> ```Go
> 
> const (
> 	OPERAND_TRANSACTION operandType = iota
> ```

Nit: Go style is camelcase for constants. For example, `operandTransaction`.
___
*[pkg/storage/engine/metamorphic/operations.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzmjacyCuUEqbyu91y-:-Lzmjacz7Z0xWMIdszUi:b-6mknge) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8bf979a8ba42a5140800da6f3a0c05114d207e/pkg/storage/engine/metamorphic/operations.go#L32)):*
> ```Go
> // dependantOps commands should be stateless, with all state stored in the
> // passed-in test runner or its operand managers.
> type mvccOp struct {
> ```

You took a different approach here than what is done in the Pebble metamorphic test. Rather than having an `op` interface with concrete implementations that refer to concrete arguments, you've abstracted the operands. Do you mind motivating this different approach? On the surface, it seems to push more work to the op execution phase. That isn't necessarily problematic. It also mixes in generation with execution (e.g. `mvccOp.weight` which shouldn't be needed at execution time). I'm also not quite clear on what to make of `dependentOps`. In the Pebble metamorphic test, op dependencies are implicit in their ordering. 


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44458)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @petermattis, and @sumeerbhola)

___
*[pkg/storage/engine/metamorphic/operations.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzmjacyCuUEqbyu91y-:-LznOgY359JK7wZBlqSl:b-wxyv64) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8bf979a8ba42a5140800da6f3a0c05114d207e/pkg/storage/engine/metamorphic/operations.go#L32)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You took a different approach here than what is done in the Pebble metamorphic test. Rather than having an `op` interface with concrete implementations that refer to concrete arguments, you've abstracted the operands. Do you mind motivating this different approach? On the surface, it seems to push more work to the op execution phase. That isn't necessarily problematic. It also mixes in generation with execution (e.g. `mvccOp.weight` which shouldn't be needed at execution time). I'm also not quite clear on what to make of `dependentOps`. In the Pebble metamorphic test, op dependencies are implicit in their ordering. 
</blockquote></details>

The operand abstraction came from an ideal goal of being able to cleanly define what operands depended on others - so if you need a batch and have zero batches, you'll just create one by inserting a batch's `opener`, and the same for transactions, etc. I figured this would be an easier way to generate correct sequences of operations, to just insert openers as necessary, instead of restricting your generator to the set of legal operations (a set that can be of high complexity to completely generate; the interaction between batches and transactions being the most tricky one that comes to mind). The hope was also that all state information about active objects would live in their managers only.

In practice it all ended up being more intertwined and less modular than I was hoping, but the general idea still works.

All non-opener dependent operations are defined in `dependentOps`. I should probably describe this in a comment too, but basically, every time you "randomly generate" an operation, you go through its operands list, see if any need to be opened, then recursively run their openers (if necessary), then run the `dependentOps` method, and recursively run those operations (which could also require running more openers and more dependentOps), then you run the operation itself. 


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44458)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @sumeerbhola)

___
*[pkg/storage/engine/metamorphic/operands.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzqkzlyDzSeIpph7u14:-LzqkzlyDzSeIpph7u15:b-g81jn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a295ae9ceaa87200c552c73ea918a6b19aa8e404/pkg/storage/engine/metamorphic/operands.go#L58)):*
> ```Go
> }
> 
> type operand interface{}
> ```

Usage of the empty interface is a big signal that something is off with the abstractions. The required use of type assertions is a related indicator. My instinct is that there should be abstraction at the op level, with concrete implementations of the various MVCC operations that know the concrete types of their parameters/operands. For example:

```
type mvccScanOp struct {
  reader Reader
  start roachpb.Key
  end roachpb.Key
  max int64
  ...
}
```

One nuance here is that you don't want to store an actual `engine.Reader` in `mvccScanOp`. That's why the Pebble metamorphic tests use IDs for all of the objects. The level of indirection allows an op to refer to an object that hasn't been created yet. This is similar to the way you have ops retrieving objects from managers.
___
*[pkg/storage/engine/metamorphic/operations.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzmjacyCuUEqbyu91y-:-LzqjA_X6BiIU5ELMjBw:b-uh044b) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8bf979a8ba42a5140800da6f3a0c05114d207e/pkg/storage/engine/metamorphic/operations.go#L32)):*
> The operand abstraction came from an ideal goal of being able to cleanly define what operands depended on others - so if you need a batch and have zero batches, you'll just create one by inserting a batch's opener, and the same for transactions, etc.

I don't think it implies the `operand` abstraction. I think you primarily get this functionality via the `mvccOp.operands` specification.

It isn't clear to me if this goal is necessary. The Pebble metamorphic test generator simply skips generating an op if it is applied to an object for which there are no live objects. For example, if you try to generate an `iter.Close` op and there are no live iterators. Immediately opening and then closing an iterator doesn't seem to be terribly useful. 

> The hope was also that all state information about active objects would live in their managers only.

I'm good with placing the state information about active objects in the managers. My feeling is that the managers should be a op-generation time component. Currently they are also an op-execution component as well. 

> In practice it all ended up being more intertwined and less modular than I was hoping, but the general idea still works.

I think it is worthwhile to disentangle the code and state needed for generation from the state needed for executing ops. The current `mvccOps` structure is good on the generation side, but the way execution is mixed in there is a bit messy. And I think this will get even messier when you add parsing and what not. 


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
This testing infrastructure is complex enough, that it is beneficial for it to have its own tests so you can debug the component pieces in isolation. It would be nice to have a test that outputs the ops that are generated (without running them) for a particular seed and count. See also `TestGenerator` in the Pebble metamorphic test which verifies that all of the live objects are properly closed. Another bit that could use a test is the parsing. Right now that is a bit difficult because execution is interwoven with parsing. Being able to test parsing in isolation is a good reason for disentangling that. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-:-LzquWta2nauUqxc65I5:b-7idq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @sumeerbhola)

___
*[pkg/storage/engine/metamorphic/generator.go, line 248 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzqsuQMA_3BdrRTWy8U:-LzqsuQMA_3BdrRTWy8V:b-p0pffc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a295ae9ceaa87200c552c73ea918a6b19aa8e404/pkg/storage/engine/metamorphic/generator.go#L248)):*
> ```Go
> func (t *tsGenerator) randomPastTimestamp(rng *rand.Rand) hlc.Timestamp {
> 	var result hlc.Timestamp
> 	result.WallTime = int64(float64(t.lastTS.WallTime+1) * rng.Float64())
> ```

I suspect it would be better for the historic times to be skewed towards the current time. Not sure precisely how I would accomplish that. Perhaps with a zipfian distribution.
___
*[pkg/storage/engine/metamorphic/meta_test.go, line 119 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzqsPl34xEqxUEHG3tz:-LzqsPl34xEqxUEHG3u-:b3qcxec) ([raw file](https://github.com/cockroachdb/cockroach/blob/a295ae9ceaa87200c552c73ea918a6b19aa8e404/pkg/storage/engine/metamorphic/meta_test.go#L119)):*
> ```Go
> 	defer leaktest.AfterTest(t)
> 	ctx := context.Background()
> 	seeds := []int64{123}
> ```

By default, I think you should generate a few random seeds. Perhaps make this a command line flag.
___
*[pkg/storage/engine/metamorphic/meta_test.go, line 136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzqsFIH0yYKZ-_XeQTD:-LzqsFIH0yYKZ-_XeQTE:bh4yxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a295ae9ceaa87200c552c73ea918a6b19aa8e404/pkg/storage/engine/metamorphic/meta_test.go#L136)):*
> ```Go
> 	}
> 	for _, seed := range seeds {
> 		t.Run(fmt.Sprintf("seed=%d", seed), func(t *testing.T) {
> ```

The seed won't generate the same series of operations if the use of the random number generator changes. Is that a problem?
___
*[pkg/storage/engine/metamorphic/operands.go, line 112 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzqtOq19Hv4OTGfUDe1:-LzqtOq19Hv4OTGfUDe2:b-qc01e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a295ae9ceaa87200c552c73ea918a6b19aa8e404/pkg/storage/engine/metamorphic/operands.go#L112)):*
> ```Go
> 	}
> 
> 	return k.liveKeys[int(k.rng.Float64()*float64(len(k.liveKeys)))]
> ```

Better to do `k.liveKeys[k.rng.Intn(len(k.liveKeys))]`. Using floats to generate random numbers in a range like this is actually not uniformly random which is sort of surprising. I don't think it really matters, but it is better to use the supported idiom for generating integers in a range.


<!-- Sent from Reviewable.io -->

Commit:
engine/metamorphic: Implement more mvcc ops, close pebble/rocks gap

Implements more MVCC operations such as MVCCDelete,
MVCCInconsistent{Get,Scan}.

Also updates batch iterator use to make the behaviour appear the
same between rocksdb and pebble. There are some minor differences
in the implementations: for instance, RocksDB batch iterators don't
support SeekLT, Prev, don't respond well to SeekGEs before lower bounds,
and appear to be SeekGE'd to the lower bound upon initialization.
This branch ensures batch iterators are used so that TestMeta
runs on rocksdb and pebble generate the same output. Important
for when output from one type of TestMeta run is --check ed with
the other engine.

Release note: None.

itsbilal commented:
Thanks for the review!

I just added support for more operations, as well as standardized behaviour between rocksdb and pebble enough that you can pass output from one engine into the other with --check and it should always pass.

Functionally I'll cap this PR off at this point. I'll make another parallel PR for the restart point work - which should be straightforward now that everything it depends on is done. Once that is up and running (hopefully with no obvious failures) I'll come back here and address code review comments - so thanks for doing a review quickly.

I'm open to pulling out all execution bits from the `mvccOp` struct and making it part of a concrete struct for each operation, complete with concrete bound parameters (that might also need to be IDs, for the reason you described.). `operands` can be ID `strings` going forward, and just parsed by the relevant manager upon execution time. The `dependentOps` stuff can be purely a generation-time factor, as well as a new `onGenerated` handler can handle populating any managers with the new state of the universe post-op generation (eg. removing closed IDs, keeping track of batches with in-flight writes for a transaction, etc).

And speaking about the file checking mode, it turned out to be relatively straightforward, because assuming the rest of the state machine behaves predictably and the output file was well generated, all `dependentOps` and openers would have already run by the time an operation shows up.

petermattis reviewed (commented):
> Functionally I'll cap this PR off at this point.

Let me give this another review today and we'll try to get it merged in its current form. The suggestions I'm making can be handled in follow-on PRs. Probably easier for everyone that way.

> operands can be ID strings going forward, and just parsed by the relevant manager upon execution time.

I'm suggesting that there is an explicit parsing step for both operators and operands. If an operator has an int operator, it should be an `int` field in the associated `mvccOp` struct.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-:-M-ANBNwD2NZRbyFZE5l:b-8h2nrh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @sumeerbhola)



<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44458)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @sumeerbhola)

___
*[pkg/storage/engine/metamorphic/deck.go, line 24 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-F_3Ma-yvXHoXai2kZ:-M-F_3Ma-yvXHoXai2k_:b1grfkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/deck.go#L24)):*
> ```Go
> // ensures that each element is returned with a desired frequency within the
> // size of the deck.
> type Deck struct {
> ```

Only used within this package, so this can be unexported: `s/Deck/deck/g`.
___
*[pkg/storage/engine/metamorphic/generator.go, line 100 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-F_Au3BIrsu3DTK_o6:-M-F_Au3BIrsu3DTK_o7:b-fv0uti) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/generator.go#L100)):*
> ```Go
> // generateAndRun generates n operations using a TPCC-style deck shuffle with
> // weighted probabilities of each operation appearing.
> func (m *metaTestRunner) generateAndRun(n uint64) {
> ```

Nit: we're unlikely to run more than 2 billion operations. `s/uint64/int/g`. This also simplifies the loop below slightly.
___
*[pkg/storage/engine/metamorphic/generator.go, line 101 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-F_cI3D2saIhTKJQkt:-M-F_cI3D2saIhTKJQku:b-7xrcku) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/generator.go#L101)):*
> ```Go
> // weighted probabilities of each operation appearing.
> func (m *metaTestRunner) generateAndRun(n uint64) {
> 	m.ops = make([]*mvccOp, n)
> ```

You're initializing `m.ops` to length `n`, but then seem to be appending to `m.ops` in `runOp`. I think the end result is that you'll end up with `len(m.ops) == 2 * n`. Perhaps you meant to `make([]*mvccOp, 0, n)`?
___
*[pkg/storage/engine/metamorphic/generator.go, line 105 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-F_OJq13F833zJx87U:-M-F_OJq13F833zJx87V:b-x5z4rc) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/generator.go#L105)):*
> ```Go
> 
> 	for i := uint64(0); i < n; i++ {
> 		opToAdd := &operations[deck.Int()]
> ```

Are you adding the op somewhere? Seems like you're just running it. Perhaps `s/opToAdd/op/g`.
___
*[pkg/storage/engine/metamorphic/meta_test.go, line 47 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-FudNh4Qc4foBSjQHP:-M-FudNh4Qc4foBSjQHQ:b-2ppj65) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/meta_test.go#L47)):*
> ```Go
> // createTestPebbleEngine returns a new in-memory Pebble storage engine.
> func createTestPebbleEngine(path string) (engine.Engine, error) {
> 	return engine.NewEngine(enginepb.EngineTypePebble, 1<<20, base.StorageConfig{
> ```

The `base.StorageConfig` looks identical between this function and `createTestRocksDBEngine`. Perhaps pull out into a helper: `func makeStorageConfig(dir string) base.StorageConfig`.
___
*[pkg/storage/engine/metamorphic/meta_test.go, line 109 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-FxRw0AQWaYBDNMH2f:-M-FxRw0AQWaYBDNMH2g:b-tylm8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/meta_test.go#L109)):*
> ```Go
> 				testRunner.parseFileAndRun(checkFile)
> 			} else {
> 				testRunner.generateAndRun(10000)
> ```

Add a TODO about making this configurable. Note that in Pebble, the default is a random number in the range `[5000,10000)`.
___
*[pkg/storage/engine/metamorphic/operands.go, line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-FxqiUDpgYlcLYhv8k:-M-FxqiUDpgYlcLYhv8l:b-8zqw42) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operands.go#L49)):*
> ```Go
> // writers, etc) should be stored in an operandManager.
> type operandManager interface {
> 	get() operand
> ```

Can you provide some commentary around these? `get()` seems to generate an operand (would `generate` be a better name?). `opener()` seems to return the name of an operator. 
___
*[pkg/storage/engine/metamorphic/operands.go, line 178 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G1pQ028UHHozzPaqw:-M-G1pQ028UHHozzPaqx:b-2sdntq) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operands.go#L178)):*
> ```Go
> 	liveTxns        []*roachpb.Transaction
> 	txnIdMap        map[string]*roachpb.Transaction
> 	inFlightBatches map[*roachpb.Transaction][]engine.Batch
> ```

`s/inFlightBatches/openBatches/g`?
___
*[pkg/storage/engine/metamorphic/operands.go, line 256 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G1jtt-AMD2i9KHD1L:-M-G1jtt-AMD2i9KHD1M:b-ax62v3) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operands.go#L256)):*
> ```Go
> 				i--
> 				found = true
> 				// Don't break; this batch could appear multiple times.
> ```

Can you elaborate? Why would the same batch appear multiple times in the in-flight batches list.
___
*[pkg/storage/engine/metamorphic/operations.go, line 131 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G2fmBEz-ZwaHBPmSN:-M-G2fmBEz-ZwaHBPmSO:b7bmibg) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operations.go#L131)):*
> ```Go
> 
> // List of operations, where each operation is defined as one instance of mvccOp.
> var operations = []mvccOp{
> ```

I think there are some operations missing here. Can you add a TODO? For example, `MVCCConditionalPut`, `MVCCBlindPut`, `MVCCMerge`, `MVCCClearTimeRange`. Perhaps you've added these in subsequent PRs.

I'm not seeing ingestion, which is another important bit of surface area to trigger. The way ingestion works in the Pebble metamorphic test is that it converts a batch into an sstable and ingests that. Something similar could be done here. Again, a TODO is sufficient for this PR.
___
*[pkg/storage/engine/metamorphic/operations.go, line 161 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G2N561_9heUhuq0i8:-M-G2N561_9heUhuq0i9:b5ph1ul) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operations.go#L161)):*
> ```Go
> 			val, intent, err := engine.MVCCGet(ctx, reader, key.Key, txn.ReadTimestamp, engine.MVCCGetOptions{
> 				Inconsistent: false,
> 				Tombstones:   true,
> ```

Add a TODO that all of these parameters should be operands. 
___
*[pkg/storage/engine/metamorphic/operations.go, line 313 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G0go6A-zR3AtMSQqz:-M-G0go6A-zR3AtMSQr-:b-hk5n51) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operations.go#L313)):*
> ```Go
> 				result = append(result, opRun{
> 					op:   m.nameToOp["batch_commit"],
> 					args: []operand{batch},
> ```

I don't want to beat a dead horse too severely, but this is another indication of the problem with the current abstractions. `"batch_commit"` is referring to an `mvccOp` and requiring specific arguments. There is no type checking here on the operands. No type checking that you've spelled `"batch_commit"` correctly. The alternative would be to have something like:

```
result = append(result, batchCommitOp{
    batch: batch,
 })
```

Let's leave this change to a follow-on PR, as it will be extensive.
___
*[pkg/storage/engine/metamorphic/operations.go, line 529 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G3Oo88F6iJcws-Nk3:-M-G3Oo88F6iJcws-Nk4:b-11yt0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operations.go#L529)):*
> ```Go
> 				tmpKey := key
> 				key = endKey
> 				endKey = tmpKey
> ```

The Go idiom for swapping two variables is: `key, endKey = endKey, key`. No need for `tmpKey`.
___
*[pkg/storage/engine/metamorphic/operations.go, line 563 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G3n0W7CQZ_mc-_wCH:-M-G3n0W7CQZ_mc-_wCI:b-mquzmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operations.go#L563)):*
> ```Go
> 			OPERAND_MVCC_KEY,
> 		},
> 		weight: 5,
> ```

I suspect the weight of this operation should be even lower than this. We don't perform manual compactions very often, usually only after a delete-range.


<!-- Sent from Reviewable.io -->

Commit:
engine/metamorphic: Address code review changes

Lots of minor changes such as renaming all operandTypes to
camel casing, TODO insertions wherever necessary, the use
of the pastTS operand type for inconsistent gets/scans, and
the use of rand.Zipf for generating past timestamps.

Release note: None

itsbilal reviewed (commented):
Thanks for the very extensive review, really appreciated! I've addressed all the simpler points, and noted the major refactor ones for a follow-up PR. I'll update the restarts PR once this is merged.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-:-M-HX2Sf2Ry10R6CKH2R:bta74wa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis and @sumeerbhola)

___
*[pkg/storage/engine/metamorphic/deck.go, line 24 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-F_3Ma-yvXHoXai2kZ:-M-GljTl8xmA6XUz8_kR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/deck.go#L24)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Only used within this package, so this can be unexported: `s/Deck/deck/g`.
</blockquote></details>

Done.
___
*[pkg/storage/engine/metamorphic/generator.go, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzmlExm6KeNUqc9bOCu:-M-GpSei4F1TuI6_-JX-:b4zmhpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8bf979a8ba42a5140800da6f3a0c05114d207e/pkg/storage/engine/metamorphic/generator.go#L35)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I suspect there is a level of abstraction here that isn't serving much purpose. Given that there are a fixed number of "*managers", rather than a map you could embed the concrete types here. There are a few places where you're using a variable as a key to `managers`, but my suspicion is that is due to the operand abstraction which I'm also questioning.
</blockquote></details>

Deferred for a followup PR that refactors operand abstraction.
___
*[pkg/storage/engine/metamorphic/generator.go, line 248 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzqsuQMA_3BdrRTWy8U:-M-GvI3e9FbyuYGWLLYp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a295ae9ceaa87200c552c73ea918a6b19aa8e404/pkg/storage/engine/metamorphic/generator.go#L248)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I suspect it would be better for the historic times to be skewed towards the current time. Not sure precisely how I would accomplish that. Perhaps with a zipfian distribution.
</blockquote></details>

Done.
___
*[pkg/storage/engine/metamorphic/generator.go, line 100 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-F_Au3BIrsu3DTK_o6:-M-Gq51KEcQBMGLQttzy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/generator.go#L100)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nit: we're unlikely to run more than 2 billion operations. `s/uint64/int/g`. This also simplifies the loop below slightly.
</blockquote></details>

Done.
___
*[pkg/storage/engine/metamorphic/generator.go, line 101 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-F_cI3D2saIhTKJQkt:-M-Gpo6e38pG2qcEJ78L:bxllgvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/generator.go#L101)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You're initializing `m.ops` to length `n`, but then seem to be appending to `m.ops` in `runOp`. I think the end result is that you'll end up with `len(m.ops) == 2 * n`. Perhaps you meant to `make([]*mvccOp, 0, n)`?
</blockquote></details>

Done. Whoops. I was keeping this around for printing output runs in the end, but it's ultimately unnecessary so I got rid of this slice altogether.
___
*[pkg/storage/engine/metamorphic/generator.go, line 105 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-F_OJq13F833zJx87U:-M-GqH29Avz9biNkRkoy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/generator.go#L105)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Are you adding the op somewhere? Seems like you're just running it. Perhaps `s/opToAdd/op/g`.
</blockquote></details>

Done.
___
*[pkg/storage/engine/metamorphic/meta_test.go, line 119 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzqsPl34xEqxUEHG3tz:-M-GxMwRErxJ4MGS6YSH:b-9mft6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/a295ae9ceaa87200c552c73ea918a6b19aa8e404/pkg/storage/engine/metamorphic/meta_test.go#L119)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

By default, I think you should generate a few random seeds. Perhaps make this a command line flag.
</blockquote></details>

Given test runtime can be quite slow, I've resorted to adding 1 fixed seed, 1 user-specified seed, and 1 random seed.
___
*[pkg/storage/engine/metamorphic/meta_test.go, line 136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzqsFIH0yYKZ-_XeQTD:-M-GxTWl33UzW0ijwxeP:bfqjrfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a295ae9ceaa87200c552c73ea918a6b19aa8e404/pkg/storage/engine/metamorphic/meta_test.go#L136)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The seed won't generate the same series of operations if the use of the random number generator changes. Is that a problem?
</blockquote></details>

It shouldn't be an issue. The purpose of having a seeded RNG is to more easily be able to recreate a test run if the output file gets lost for whatever reason (eg. on CI). Presumably if there's a code change, the seed will not exactly match up to the same deterministic set of operations, but that's alright.
___
*[pkg/storage/engine/metamorphic/meta_test.go, line 47 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-FudNh4Qc4foBSjQHP:-M-GwBmm6zkUwVCEXFpf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/meta_test.go#L47)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The `base.StorageConfig` looks identical between this function and `createTestRocksDBEngine`. Perhaps pull out into a helper: `func makeStorageConfig(dir string) base.StorageConfig`.
</blockquote></details>

Done.
___
*[pkg/storage/engine/metamorphic/meta_test.go, line 109 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-FxRw0AQWaYBDNMH2f:-M-HWzfIEBvgAz_6ez2s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/meta_test.go#L109)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Add a TODO about making this configurable. Note that in Pebble, the default is a random number in the range `[5000,10000)`.
</blockquote></details>

Done.
___
*[pkg/storage/engine/metamorphic/operands.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-Lzmlz4zEP89aGWCO0vI:-M-HG2UuBzKvKtq4cicG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8bf979a8ba42a5140800da6f3a0c05114d207e/pkg/storage/engine/metamorphic/operands.go#L30)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nit: Go style is camelcase for constants. For example, `operandTransaction`.
</blockquote></details>

Done.
___
*[pkg/storage/engine/metamorphic/operands.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzqkzlyDzSeIpph7u14:-M-HMzoQ7jfUqlODRtLB:b-clrg9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a295ae9ceaa87200c552c73ea918a6b19aa8e404/pkg/storage/engine/metamorphic/operands.go#L58)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Usage of the empty interface is a big signal that something is off with the abstractions. The required use of type assertions is a related indicator. My instinct is that there should be abstraction at the op level, with concrete implementations of the various MVCC operations that know the concrete types of their parameters/operands. For example:

```
type mvccScanOp struct {
  reader Reader
  start roachpb.Key
  end roachpb.Key
  max int64
  ...
}
```

One nuance here is that you don't want to store an actual `engine.Reader` in `mvccScanOp`. That's why the Pebble metamorphic tests use IDs for all of the objects. The level of indirection allows an op to refer to an object that hasn't been created yet. This is similar to the way you have ops retrieving objects from managers.
</blockquote></details>

Noted. Will address this in a refactor in an upcoming PR.
___
*[pkg/storage/engine/metamorphic/operands.go, line 112 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzqtOq19Hv4OTGfUDe1:-M-HOAP_D_H0JGfMAqTg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a295ae9ceaa87200c552c73ea918a6b19aa8e404/pkg/storage/engine/metamorphic/operands.go#L112)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Better to do `k.liveKeys[k.rng.Intn(len(k.liveKeys))]`. Using floats to generate random numbers in a range like this is actually not uniformly random which is sort of surprising. I don't think it really matters, but it is better to use the supported idiom for generating integers in a range.
</blockquote></details>

Done.
___
*[pkg/storage/engine/metamorphic/operands.go, line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-FxqiUDpgYlcLYhv8k:-M-HMMeFCnIBOaIpjrjQ:b-1m9nyn) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operands.go#L49)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Can you provide some commentary around these? `get()` seems to generate an operand (would `generate` be a better name?). `opener()` seems to return the name of an operator. 
</blockquote></details>

Done. Get usually does not generate an operand if there already exists one. I say usually because keys are an exception; we want to sometimes reuse keys and other times return new ones. Added commentary about this.
___
*[pkg/storage/engine/metamorphic/operands.go, line 178 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G1pQ028UHHozzPaqw:-M-HODQC7CXMFdb5p7mQ:bseo6br) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operands.go#L178)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/inFlightBatches/openBatches/g`?
</blockquote></details>

It's batches that have writes that need to happen before this transaction commits - so maybe "in-flight" makes more sense? The transaction struct itself has an inFlightWrites field.
___
*[pkg/storage/engine/metamorphic/operands.go, line 256 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G1jtt-AMD2i9KHD1L:-M-HOTkd28ADc-A9vcJP:b-emmkxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operands.go#L256)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Can you elaborate? Why would the same batch appear multiple times in the in-flight batches list.
</blockquote></details>

At write time, we don't check if this transaction has already written to this batch. We just append it to the end of this slice. It's up to the read case to resolve the dedups. I've clarified it here.
___
*[pkg/storage/engine/metamorphic/operations.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzmjacyCuUEqbyu91y-:-M-HTw8KDGsw9CtxXVEl:b-os46tg) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8bf979a8ba42a5140800da6f3a0c05114d207e/pkg/storage/engine/metamorphic/operations.go#L32)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

> The operand abstraction came from an ideal goal of being able to cleanly define what operands depended on others - so if you need a batch and have zero batches, you'll just create one by inserting a batch's opener, and the same for transactions, etc.

I don't think it implies the `operand` abstraction. I think you primarily get this functionality via the `mvccOp.operands` specification.

It isn't clear to me if this goal is necessary. The Pebble metamorphic test generator simply skips generating an op if it is applied to an object for which there are no live objects. For example, if you try to generate an `iter.Close` op and there are no live iterators. Immediately opening and then closing an iterator doesn't seem to be terribly useful. 

> The hope was also that all state information about active objects would live in their managers only.

I'm good with placing the state information about active objects in the managers. My feeling is that the managers should be a op-generation time component. Currently they are also an op-execution component as well. 

> In practice it all ended up being more intertwined and less modular than I was hoping, but the general idea still works.

I think it is worthwhile to disentangle the code and state needed for generation from the state needed for executing ops. The current `mvccOps` structure is good on the generation side, but the way execution is mixed in there is a bit messy. And I think this will get even messier when you add parsing and what not. 
</blockquote></details>

Noted. Will address in a follow-up PR. 
___
*[pkg/storage/engine/metamorphic/operations.go, line 131 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G2fmBEz-ZwaHBPmSN:-M-HUOlrDtBCzVZsIhx5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operations.go#L131)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think there are some operations missing here. Can you add a TODO? For example, `MVCCConditionalPut`, `MVCCBlindPut`, `MVCCMerge`, `MVCCClearTimeRange`. Perhaps you've added these in subsequent PRs.

I'm not seeing ingestion, which is another important bit of surface area to trigger. The way ingestion works in the Pebble metamorphic test is that it converts a batch into an sstable and ingests that. Something similar could be done here. Again, a TODO is sufficient for this PR.
</blockquote></details>

Done.
___
*[pkg/storage/engine/metamorphic/operations.go, line 161 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G2N561_9heUhuq0i8:-M-HVzCy-EVKbLb9v2o0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operations.go#L161)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Add a TODO that all of these parameters should be operands. 
</blockquote></details>

Done.
___
*[pkg/storage/engine/metamorphic/operations.go, line 313 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G0go6A-zR3AtMSQqz:-M-HWZ89BoS33iSt6oGl:bc8tnxl) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operations.go#L313)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I don't want to beat a dead horse too severely, but this is another indication of the problem with the current abstractions. `"batch_commit"` is referring to an `mvccOp` and requiring specific arguments. There is no type checking here on the operands. No type checking that you've spelled `"batch_commit"` correctly. The alternative would be to have something like:

```
result = append(result, batchCommitOp{
    batch: batch,
 })
```

Let's leave this change to a follow-on PR, as it will be extensive.
</blockquote></details>

Noted for a follow on PR. Thanks!
___
*[pkg/storage/engine/metamorphic/operations.go, line 529 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G3Oo88F6iJcws-Nk3:-M-HWjJy5J0q-hRAi_6p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operations.go#L529)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The Go idiom for swapping two variables is: `key, endKey = endKey, key`. No need for `tmpKey`.
</blockquote></details>

Done.
___
*[pkg/storage/engine/metamorphic/operations.go, line 563 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G3n0W7CQZ_mc-_wCH:-M-HWpauCghFmidi2eDF:b-mvwub5) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operations.go#L563)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I suspect the weight of this operation should be even lower than this. We don't perform manual compactions very often, usually only after a delete-range.
</blockquote></details>

Lowered it to 2. 


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo the comment on the type of `inFlightBatches`. Definitely want to see the follow-on PRs as I think the clarification will be significant.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-:-M-M2PBd-FJjsG0gaK2U:b-hk4pah)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal and @sumeerbhola)

___
*[pkg/storage/engine/metamorphic/generator.go, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzmlExm6KeNUqc9bOCu:-M-M-orFBC1Gq-hutUSy:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8bf979a8ba42a5140800da6f3a0c05114d207e/pkg/storage/engine/metamorphic/generator.go#L35)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Deferred for a followup PR that refactors operand abstraction.
</blockquote></details>

Ack.
___
*[pkg/storage/engine/metamorphic/operands.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzqkzlyDzSeIpph7u14:-M-M20E4-tS3ps9O2H1W:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a295ae9ceaa87200c552c73ea918a6b19aa8e404/pkg/storage/engine/metamorphic/operands.go#L58)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Noted. Will address this in a refactor in an upcoming PR.
</blockquote></details>

Ack.
___
*[pkg/storage/engine/metamorphic/operands.go, line 178 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G1pQ028UHHozzPaqw:-M-M0v17-XKje0apO7wB:b-o4hqfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operands.go#L178)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

It's batches that have writes that need to happen before this transaction commits - so maybe "in-flight" makes more sense? The transaction struct itself has an inFlightWrites field.
</blockquote></details>

The transaction's in-flight writes refers to network operations (I think). 
___
*[pkg/storage/engine/metamorphic/operands.go, line 256 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G1jtt-AMD2i9KHD1L:-M-M1N6X2yJ1ROw8NWiN:bienj7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operands.go#L256)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

At write time, we don't check if this transaction has already written to this batch. We just append it to the end of this slice. It's up to the read case to resolve the dedups. I've clarified it here.
</blockquote></details>

Seems like what you want is the set of uncommitted batches per transaction. Could you change the type of `inFlightBatches` to `map[*roachpb.Transaction]map[engine.Batch]struct{}`?
___
*[pkg/storage/engine/metamorphic/operations.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-LzmjacyCuUEqbyu91y-:-M-M1_ow0oljXr-3wwXp:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8bf979a8ba42a5140800da6f3a0c05114d207e/pkg/storage/engine/metamorphic/operations.go#L32)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Noted. Will address in a follow-up PR. 
</blockquote></details>

Ack.
___
*[pkg/storage/engine/metamorphic/operations.go, line 131 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G2fmBEz-ZwaHBPmSN:-M-M1dlM9vEyY3PUZiQA:b9kmcel) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operations.go#L131)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I'm not sure if my list was exhaustive. Did you double-check that there aren't other operations?
___
*[pkg/storage/engine/metamorphic/operations.go, line 313 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G0go6A-zR3AtMSQqz:-M-M1nlL7_TSVwi_qxV_:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operations.go#L313)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Noted for a follow on PR. Thanks!
</blockquote></details>

Ack.


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
Thanks for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-:-M-MjMpa8TJznuYtOmJo:b-9qdoz3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @petermattis, and @sumeerbhola)

___
*[pkg/storage/engine/metamorphic/operands.go, line 178 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G1pQ028UHHozzPaqw:-M-MjM-zABp5-NkgPHgS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operands.go#L178)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The transaction's in-flight writes refers to network operations (I think). 
</blockquote></details>

Done.
___
*[pkg/storage/engine/metamorphic/operands.go, line 256 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G1jtt-AMD2i9KHD1L:-M-MhlY83IPHdJOBPIF_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operands.go#L256)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Seems like what you want is the set of uncommitted batches per transaction. Could you change the type of `inFlightBatches` to `map[*roachpb.Transaction]map[engine.Batch]struct{}`?
</blockquote></details>

Done.
___
*[pkg/storage/engine/metamorphic/operations.go, line 131 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-G2fmBEz-ZwaHBPmSN:-M-MiPrQ0bQEaLAdGgXe:bji1pve) ([raw file](https://github.com/cockroachdb/cockroach/blob/20103e1166eed359daaaf75b7f73f94b0dc6b5cb/pkg/storage/engine/metamorphic/operations.go#L131)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm not sure if my list was exhaustive. Did you double-check that there aren't other operations?
</blockquote></details>

There are others: MVCCIterate (seems unused), MVCCResolveWriteIntent (which we kinda already call in txn_commit), MVCCFindSplitKey, MVCCIncrement, MVCCInitPut, as well as some variants of Get/Put meant for protos that we can probably ignore. I'll add them too, and say that this isn't an exhaustive list.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44458)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @sumeerbhola)

___
*[pkg/storage/engine/metamorphic/operands.go, line 262 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-NTDI58-534c2rzjNJ:-M-NTDI58-534c2rzjNK:b-4llvbs) ([raw file](https://github.com/cockroachdb/cockroach/blob/1659148cac89d6e3fe59a77017050b7681cbb7c0/pkg/storage/engine/metamorphic/operands.go#L262)):*
> ```Go
> func (t *txnManager) clearBatch(batch engine.Batch) {
> 	for _, batches := range t.openBatches {
> 		for batch2 := range batches {
> ```

Is this loop necessary? Can't you blindly `delete(batches, batch)` here?


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
Addressed linter-reported issues too. TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-:-M-QEHnP3qv1ROA1nd57:b-p1lma9)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @petermattis, and @sumeerbhola)

___
*[pkg/storage/engine/metamorphic/operands.go, line 262 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/44458#-M-NTDI58-534c2rzjNJ:-M-Q9VYfDEkNxuC7qCU9:b7e7zy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1659148cac89d6e3fe59a77017050b7681cbb7c0/pkg/storage/engine/metamorphic/operands.go#L262)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is this loop necessary? Can't you blindly `delete(batches, batch)` here?
</blockquote></details>

:man_facepalming: yes, you can. My bad.


<!-- Sent from Reviewable.io -->

itsbilal commented:
bors r+
